src/mapblock.cpp

```diff
@@ -494,15 +524,20 @@ void MapBlock::deSerialize(std::istream &is, u8 version, bool disk)
 			<<": Node metadata"<<std::endl);
 	// Ignore errors
 	try {
-		std::ostringstream oss(std::ios_base::binary);
-		decompressZlib(is, oss);
-		std::istringstream iss(oss.str(), std::ios_base::binary);
-		if (version >= 23)
-			m_node_metadata.deSerialize(iss, m_gamedef->idef());
-		else
-			content_nodemeta_deserialize_legacy(iss,
-				&m_node_metadata, &m_node_timers,
-				m_gamedef->idef());
+		if (version >= 29) {
+			m_node_metadata.deSerialize(is, m_gamedef->idef());
+		} else {
+			// resuse in_raw
+			in_raw.str("");
+			in_raw.clear();
+			decompress(is, in_raw, version);
+			if (version >= 23)
+				m_node_metadata.deSerialize(in_raw, m_gamedef->idef());
+			else
+				content_nodemeta_deserialize_legacy(in_raw,
+					&m_node_metadata, &m_node_timers,
+					m_gamedef->idef());
+		}
 	} catch(SerializationError &e) {
 		warningstream<<"MapBlock::deSerialize(): Ignoring an error"
 				<<" while deserializing node metadata at ("
```

Comment:
Btw this has to be made fatal instead of a warning on `ver >= 29` too because of what I mentioned.

Reply:
How about the latest? Can't say I like it, though. It creates yet another ostringstream each time we safe a mapblock and there are *two* additional copy of the bytes representing the metadata.

Reply:
Fine by me.